### ISSUE TITLE
[bug] The reranker makes retrieval worse at every k

### ISSUE THREAD

**Issue by akash-coded** [akash-coded/nanorag#4]
### Symptom

The stage-2 reranker consistently *reduced* recall relative to the fusion it was reranking:

```
k=5 rerank=none    ER=0.773  FCR=0.455
k=5 rerank=cross   ER=0.630  FCR=0.386   ← worse
k=8 rerank=none    ER=0.849  FCR=0.614
k=8 rerank=cross   ER=0.752  FCR=0.523   ← worse
```

Shipping this would have taught the opposite of the lesson: the deck's own matrix names a
cross-encoder as the default stage-2 choice and the highest quality per unit of engineering
effort.

### Investigation

1. **First hypothesis — bad weights.** Grid-searched all four hand-set coefficients. The best
   grid point still lost to no-reranking.
2. **Second hypothesis — length bias.** Added BM25-style length normalisation to the coverage
   feature. Helped, still lost.
3. **Root cause.** The scorer used *lexical features only*, so reranking a hybrid candidate
   list threw away the dense signal entirely. It was a worse BM25 applied on top of a
   well-fused list.

### Fix

Two changes:

- Add genuinely pair-wise **semantic** features — `maxsim` (each query token against the best
  matching passage token) and `doc_cosine`. `maxsim` is the one a bi-encoder structurally
  cannot produce, because it compresses the passage before it has seen the query.
- Stop hand-tuning. **Fit** the weights by logistic regression on the dev slice, with class
  weighting because gold pairs are rare.

### After

```
weighted α=0.2               ER=0.6832  FCR=0.4155
weighted α=0.2 + learned     ER=0.7645  FCR=0.4686
paired bootstrap, evidence recall: +0.0813 [+0.0475, +0.1163] → real
frozen slice: FCR 0.419 → 0.548
```

### What this changed about the curriculum

It made the toolkit more honest about something the deck only implies: **a reranker is a
model.** It has training data, it can overfit, and its gain has to survive on a slice it never
saw. That is now notebook 04 §4.10 rather than an assumption. See
[ADR-0005](../blob/main/docs/adr/0005-learned-reranker.md).

**Comment by akash-coded** [comment]
> **Priya** · cohort peer

The investigation log is the valuable part here — specifically that hypotheses 1 and 2 were both *reasonable* and both wrong, and that the grid search over weights came back losing.

That is the bit people skip when they write these up afterwards. "Grid-searched all four coefficients, best point still lost" is what tells you the problem is not the weights, and it is only convincing because it was done before the real cause was known.

Question: given the root cause was lexical-only features, was the grid search wasted effort or was it necessary to rule out the alternative?

**Comment by akash-coded** [comment]
Necessary, and I would do it again in the same order.

Had the grid search found a winning configuration, the conclusion would have been "the weights were badly set" and the feature set would never have been questioned. **The search failing is what promoted the hypothesis from "weights" to "features".** Roughly forty minutes to eliminate the cheap explanation before touching the expensive one.

The general shape, which is worth carrying: *rule out configuration before you suspect design.* Configuration is cheap to test and is usually the answer. When it is not the answer, the fact that it is not is itself strong evidence.

The finding that came out of this is now one of the three in the README, and [#27](https://github.com/akash-coded/nanorag/discussions/27) states it more sharply than the issue does:

> a reranker only helps if it can see something the first stage could not. A reranker over the same signals as the retriever is an expensive identity function.

**Regression test:** `tests/test_retrieval.py` asserts the reranked list beats the unreranked one on evidence recall at `k=8`, so a future feature-set change that reintroduces this cannot pass silently.
